### ISSUE TITLE
StreamCorruptedExecption when try to deserialize a unsafe EncryptedNumber using Java serializer

### ISSUE THREAD

**Issue by tiantianwdy** [n1analytics/javallier#60]
When default Java serializer tries to serialise a EncryptedNumber it throws a `java.io.StreamCorruptedException` :

```
java.io.StreamCorruptedException: invalid type code: 00

	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1381)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2018)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1942)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1808)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1353)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:373)
```

After tested, this is because, current code overwrote the `writeObject` method of EncryptedNumber:

```
  private void writeObject(ObjectOutputStream out) throws IOException  {
    if (this.isSafe) {
      out.defaultWriteObject();
    } else {
      out.writeObject(obfuscate());
    }
  }
```

 but didn't provide a corresponding `readObject` method.

**Comment by tiantianwdy** [comment]
Solution: 
- Add a flag to differentiate default and non-default serialization.
- Add corresponding `readObject` method.

```
  private void writeObject(ObjectOutputStream out) throws IOException  {
    if (this.isSafe) {
      out.writeInt(0);
      out.defaultWriteObject();
    } else {
      out.writeInt(1);
      out.writeObject(obfuscate());
    }
  }

  private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
    int isSafe = in.readInt();
    if(isSafe == 0){
      in.defaultReadObject();
    } else {
      in.readObject();
    }
  }

```

**Comment by mhsiah** [comment]
Could you try the following implementation and tell me if it works.
1) Delete your `readObject` method completely.
2) Replace `writeObject` with
```java
private void writeObject(ObjectOutputStream out) throws IOException  {
    if (this.isSafe) {
        out.defaultWriteObject();
    } else {
        this.obfuscate().writeObject();
    }
}
```
If `this.isSafe` is `false`, then we use `obfuscate()` to create a new `EncryptedNumber` on which we then call `writeObject()` which will then check `isSafe` on the new `EncryptedNumber` which will be true so `defaultWriteObject()` will be called. This way we shouldn't need `readObject()` at all. Seem reasonable?

**Comment by tiantianwdy** [comment]
@mhsiah The code seems working. Will do some tests on distributed environment.  Just one question,  in the unit test how should I check the equality of the `ciphertext` between the original unsafe and the obfuscated numbers. Because the value is not equal when i use the reflection method but it is equal if I use your above code.

**Comment by mhsiah** [comment]
The values should still be not equal. The problem was probably with the line `this.obfuscate().writeObject()` above which should be `this.obfuscate().writeObject(out)` (I've edited the comment above to include this fix).

**Comment by tiantianwdy** [comment]
I actually used `this.obfuscate().writeObject(out)` in the code. But the value is equal to the original one and the field `isSafe` remains `false`. I feel the serialization only works on the fields of the instance that being directly called. Not sure why... :(

**Comment by mhsiah** [comment]
Resolved by #57.
